pkg/apis/messaging/v1alpha1/channel_lifecycle.go

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2019 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	eventingduck "github.com/knative/eventing/pkg/apis/duck/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/apis/duck/v1alpha1"
+)
+
+var chCondSet = apis.NewLivingConditionSet(ChannelConditionBackingChannelReady, ChannelConditionAddressable)
+
+const (
+	// ChannelConditionReady has status True when all subconditions below have been set to True.
+	ChannelConditionReady = apis.ConditionReady
+
+	// ChannelConditionBackingChannelReady has status True when the backing Channel CRD is ready.
+	ChannelConditionBackingChannelReady apis.ConditionType = "BackingChannelReady"
+
+	// ChannelConditionAddressable has status true when this Channel meets
+	// the Addressable contract and has a non-empty hostname.
+	ChannelConditionAddressable apis.ConditionType = "Addressable"
+)
+
+// GetCondition returns the condition currently associated with the given type, or nil.
+func (cs *ChannelStatus) GetCondition(t apis.ConditionType) *apis.Condition {
+	return chCondSet.Manage(cs).GetCondition(t)
+}
+
+// IsReady returns true if the resource is ready overall.
+func (cs *ChannelStatus) IsReady() bool {
+	return chCondSet.Manage(cs).IsHappy()
+}
+
+// InitializeConditions sets relevant unset conditions to Unknown state.
+func (cs *ChannelStatus) InitializeConditions() {
+	chCondSet.Manage(cs).InitializeConditions()
+}
+
+func (cs *ChannelStatus) SetAddress(address *v1alpha1.Addressable) {
+	if cs.Address == nil {
+		cs.Address = &v1alpha1.Addressable{}
+	}
+	if address != nil && address.URL != nil {
+		cs.Address.Hostname = address.URL.Host
```

Comment:
This looks dangerous. We can only set Hostname if nothing in the URL is needed (i.e. path, query string, etc. are both empty).

Reply:
copy/paste from Matt's changes when he migrated to the new Addressable. We are doing this all over...
You'd rather not set anything in the hostname? I'm more inclined to leave it as is, and maybe then revisit how we are doing it everywhere...